### PR TITLE
Bringing back flavor text

### PR DESCRIPTION
**Pull request by CodenameB**
This PR will bring back flavor text, reverting it's deletion some time ago.

Additionally, this adds a dna.flavor_text variable that has been used to fix changelings, cloning, and DNA modifiers to allow them to copy flavor text correctly. 

Another feature makes it so that your flavor text will not appear if your face isn't recognizable. 

These two fixes should prevent all of the metagaming problems that the previous incarnation of flavor text had. 

This is my first PR in a long time, so I may be doing this wrong. There may be ways to simplify my code, I don't really understand byond that well yet.

Thanks to DamianQ on discord for the help in bugfixes.

:cl:
 * rscadd: Flavor text re-enabled. Now, changelings and genetics can copy flavor text. Flavor text will not be visible if anything would cause your face to be unrecognizable.

### CODE REVIEW

**Comment by ghost** [comment]
QUESTION, does it still show if you scramble your DNA?

**Comment by Shadowmech88** [comment]
Is the changing of flavor text logged? I remember in the past there was an issue with people using flavor text for unlogged communication.

**Comment by CodenameB** [comment]
Commandersand :
I haven't actually tested scrambling DNA, I didn't think to. I don't believe it will be an issue, since if it did remove flavor text, everyone would lose their flavor text upon getting an SE injection, or accidentally getting radiation, so that would be a problem. It WILL remove their flavor text if they copy a monkeyman's UE.

Shadowmech88:
I don't remember hearing that issue. In any case, I think I made sure not to reimport changing flavor text mid-round, so that shouldn't be a problem either.

**Comment by Shadowmech88** [comment]
The format for directly replying to people on Github is @username, it'll give them a notification even if they're not subscribed to the thread.

**Comment by Exxion** [comment]
Optimally, there would be separate flavor texts applied depending on whether you can see the face or just the body, whether you've heard them speak, etc.
But that's probably excessive for this server
Well, maybe the first example isn't

But this removes all the really bad problems with the old system, so it's fine

**Comment by Orangerush** [comment]
Imagine being the c*ders to remove flavor text in the first place instead of fixing it

**Comment by Riekland** [comment]
h*ll yeah

**Comment by cadaverouscoyote** [comment]
about time

**Comment by BobdaBiscuit** [comment]
\>just deleting the database migration file and leaving a missing number

Kek

**Comment by clusterfack** [comment]
How is this supposed to fix the problem with voice changers?

**Comment by CodenameB** [comment]
@clusterfack the voice changer mask is a HIDEFACE mask right? If so, wearing it will make it so that your flavor text isn't displayed.

**Comment by Uqualm** [comment]
https://en.wikipedia.org/wiki/Kill_Bill:_Volume_1

**Comment by ghost** [comment]
>literal who fixes flavor text when the c*ders just outright removed it

:thinking:

**Comment by PJB3005** [comment]
You literally cannot fix flavor text properly.

**Comment by PJB3005** [comment]
Unless you're planning to make flavor text fool proof bug and exploit free (good fucking luck), this shit aint getting merged on my watch mate.

**Comment by PJB3005** [comment]
Actually nevermind I decided to read the OP.

I still think it's a retarded mechanic though.

**Comment by ghost** [comment]
>PJB says he won't merge and that it'll never be properly fixed before reading the OP

:thinking:

**Comment by Orangerush** [comment]
I came here to laugh at PJB

**Comment by Riekland** [comment]
I refuse to believe that PJB is a real human being and that he, a teenage dutch girl, is allowed to make decisions or contributions to this serb

**Comment by Kammerjunk** [comment]
I also came here to laugh at PJB
Ha

**Comment by BobdaBiscuit** [comment]
Heh

**Comment by Exxion** [comment]
I still think there should be two separate parts for body and for face (for example, "he's tall" would always show, but "he's ugly" would only show if the face is visible and not fucked up), but I'm not going to insist on it

**Comment by cadaverouscoyote** [comment]
> Actually nevermind I decided to read the OP.

rotoscoping

**Comment by Shadowmech88** [comment]
>I still think there should be two separate parts for body and for face

That would require administrating flavor text, though.
Otherwise what's to stop people from putting their name in their body text so it shows up even when their face is hidden?

**Comment by Exxion** [comment]
That would be blatant metagaming and would fall firmly within the realm of illegal
Old flavor text had a strictly worse version of the same exploit, and I only ever saw one person do that. And they're permabanned now.

**Comment by CodenameB** [comment]
I put back the migration. Can I get another review?

File: vgstation13.dme [hunk 1]
```diff
@@ -1317,7 +1317,6 @@
 #include "code\modules\migrations\SS13_Prefs\004-add-space-dust.dm"
 #include "code\modules\migrations\SS13_Prefs\005-add-parallax-speed.dm"
 #include "code\modules\migrations\SS13_Prefs\006-add-tooltips.dm"
-#include "code\modules\migrations\SS13_Prefs\007-remove-flavortext.dm"
```
Comment (DamianX) [inline vgstation13.dme]: don't remove this

File: code/modules/migrations/SS13_Prefs/007-remove-flavortext.dm [hunk 2]
```diff
@@ -1,14 +1,13 @@
-/datum/migration/sqlite/ss13_prefs/_007
-	id = 7
-	name = "Remove flavor text"
+/datum/migration/sqlite/ss13_prefs/_009
```
Comment (DamianX) [inline code/modules/migrations/SS13_Prefs/007-remove-flavortext.dm]: You don't need to change this file at all, just revert it back to what it originally was.

**Review by DamianX — CHANGES_REQUESTED** [review]

File: code/modules/migrations/SS13_Prefs/007-remove-flavortext.dm [hunk 3]
```diff
@@ -11,4 +11,4 @@
 /datum/migration/sqlite/ss13_prefs/_007/down()
 	if(hasColumn("client","flavor_text"))
 		return execute("ALTER TABLE `client` ADD COLUMN flavor_text TEXT")
-	return TRUE
+return TRUE
```
Comment (DamianX) [inline code/modules/migrations/SS13_Prefs/007-remove-flavortext.dm]: You missed an indent here.

File: code/modules/mob/mob.dm [hunk 4]
```diff
@@ -1089,6 +1089,40 @@ var/list/slot_equipment_priority = list( \
 	face_atom(I)
 	I.verb_pickup(src)
 
+/mob/proc/update_flavor_text()
```
Comment (DamianX) [inline code/modules/mob/mob.dm]: This proc and the one below are useless, get rid of them.

File: code/game/atoms.dm [hunk 5]
```diff
@@ -603,6 +603,7 @@ its easier to just keep the beam vertical.
 			if(!istype(H.dna, /datum/dna))
 				H.dna = new /datum/dna(null)
 				H.dna.real_name = H.real_name
+				H.dna.flavor_text = H.dna.flavor_text
```
Comment (DamianX) [inline code/game/atoms.dm]: did you mean `H.dna.flavor_text = H.flavor_text`?

File: code/game/dna/dna_modifier.dm [hunk 6]
```diff
@@ -972,6 +974,7 @@
 				databuf.dna = src.connected.occupant.dna.Clone()
 				if(ishuman(connected.occupant))
 					databuf.dna.real_name=connected.occupant.name
+					databuf.dna.flavor_text=connected.occupant.flavor_text
```
Comment (DamianX) [inline code/game/dna/dna_modifier.dm]: I don't think copying Structural Enzymes should copy the flavor text as well?

**Comment by DamianX** [comment]
You reset pox raiders' flavor text but not nuke ops'. Did you look into this?

**Comment by CodenameB** [comment]
I very briefly thought about nuke ops, however unlike Vox they keep their original bodies, so keeping their flavor text seemed fine. Besides, their faces are covered anyway, so flavor text won't be displayed unless they are captured and stripped, in which case it is rather unimportant.

For stealth ops, a bunch of people with the same last name and no flavor text would make them an instant giveaway to metagamers. If you allow flavor text, you can never be 100% sure, even if you have flavor text memorized, that your favorite static name didn't just change names.

File: code/game/dna/dna_modifier.dm (hunk 6 above)
Comment (CodenameB) [reply code/game/dna/dna_modifier.dm]: Hm, I removed line 977 from dna_modifer.dm, so I don't know why this isn't listed as outdated.

**Comment by SonixApache** [comment]
i have a suggestion on how to fix this for the angry traditionalists that stalk the github

"Add a toggle."

**Comment by DamianX** [comment]
Did that stop the attack animation shitposting? Riddle me this.

**Comment by SonixApache** [comment]
You can't kill the shitposting, but the toggle got the feature through.

**Comment by Shadowmech88** [comment]
Does it appear as though this feature won't go through?
The votes are overwhelmingly in its favor.

**Comment by SonixApache** [comment]
¯\_(ツ)_/¯

**Comment by Exxion** [comment]
Personally, I just like to give plenty of time for review. The last changes were only a few hours ago.

**Comment by CodenameB** [comment]
I haven't seen any colabs actively against allowing it in, just PJB saying he dislikes the mechanic.

**Comment by ghost** [comment]
Don't think this fixes the problem with wizards still keeping the flavor text from your (((staticname))). Distinctly remember a round months ago where a human wizard still had Sawbone's flavor text.

Furthermore, doesn't flavortext still allow metagay in the sense that having flavortext more-or-less metaconfirms you're not a wizard, newcop or pox raider? There will ALWAYS be some problem with this dumb feature.

**Comment by cadaverouscoyote** [comment]
>having flavortext more-or-less metaconfirms you're not a wizard, newcop or pox raider

if it can still be edited mid-round, no

**Comment by ghost** [comment]
Idea: Flavortext is toggleable, and if not customized defaults to the randomized flavortext from april fools

**Comment by FudgePucker** [comment]
Big problem with this.
It encourages bay style RP where you are roleplaying by being an OC DO NOT STEAL instead of roleplaying via your actions.

**Comment by Exxion** [comment]
There is exactly no reason one would hinder the other

**Comment by nuklearcellphoneg** [comment]
Flavor text is 100% fine as a concept, the only issue that ever came from it was it being abused to see through disguises, to communicate, y'know exploits. I think having it as a 'toggle' thing is dumb too, just have it on all the time- in regards to attack animations, they never should have been added in the first place- 66 to 63 votes.

Does setting force random name/appearance remove the flavor text? If not, it needs to, otherwise it renders the point of that null and void.

**Comment by CodenameB** [comment]
@nuklearcellphoneg 
This fixes those exploits.

I'll check into appearance bans. but it is my understand that appearance ban is very rarely used anyway.

**Comment by Exxion** [comment]
The issue @Unknown-as-Captain brought up is still relevant though

**Comment by ghost** [comment]
And if this gets added can we please have the administration do something about "flavor texts" that aren't even IC descriptions but forum-signature-tier random shit that has nothing to do with the character
Case in point, Sawbone's flavor text

**Comment by CodenameB** [comment]
@Exxion I don't see what issue he mentions? Nuke ops and other human antags still get flavor text, so there is no meta. Likewise, "The wizard had Sawbone's flavor text" can in no way be used as a meta tool, other than you know who the player would be any other round. This knowledge does not benefit you. After all, you get the exact same knowledge about every single other player too.

@Unknown-as-Captain I agree, but that is an administration issue. PRs won't fix that.

**Comment by ghost** [comment]
You see (((Staticname Lesbian)))'s flavor text on a random assistant
This random assistant is not (((Staticname Lesbian)))
There is a 99% chance this means they are a wizard in disguise, now the roundtype was spoiled to me by some stupid feature in a way I couldn't avoid noticing.
It's like some faggot ickocking that you have to ignore. My gameplay is worsened by having to ignore this ickocking faggot and play dumb until I confirm the wizard IC.

**Comment by CodenameB** [comment]
@Unknown-as-Captain The obvious simple solution is don't be a stealth wizard if you're gonna be a flavor text snowflake. 

However, to actually fix it: someone can write a very quick fix (probably two lines of code) that gives wizards a second popup after they pick their name to give them the option to change their flavor text at round start. If this still isn't merged by Monday I'll do it myself. Nuke ops too, would be just as easy.

**Comment by ghost** [comment]
The obvious simple solution is let's stop adding more issues into the game and say "I'll fix it later maybe".

**Comment by CodenameB** [comment]
@Unknown-as-Captain Feel free to learn to code anytime friendo. I just don't have time to fix it at midnight on a Friday.

**Comment by Exxion** [comment]
Raiders at least *do* lose their flavor text, unless I'm fucking stupid and misread the code

**Comment by CodenameB** [comment]
Vox raiders are set to loose flavor text. If I am changing wiz and ops, then I'll probably change them too.

**Comment by Exxion** [comment]
Ops and raiders specifically are kind of problematic, because if they're trying for stealth, they could be fucked over either by their flavor text *or* their lack of flavor text, and I'm not really sure how to fix that.
I guess that applies to wizards too, but fuck stealth wizards

EDIT: Except you actually gave an at least reasonably good way a few comments above, and I just didn't see it
Carry on then

**Comment by SonixApache** [comment]
@Intigracy when

**Comment by Exxion** [comment]
When the exploit on wizards/raiders/ops is fixed
The author hasn't done anything recently though
Someone else can write those popups, or, even better, come up with a better solution
But the popups are sufficient if nobody can think of anything better

**Comment by cadaverouscoyote** [comment]
if flavor text can still be edited the same as i remember it, there's no need. when you clicked the button it brought up an empty window, and if you didn't re-enter your flavor text yourself then it would be cleared. to my knowledge there's no way to discern if somebody accidentally cleared it or not.

but maybe i just don't fuckin get it because i never even saw the problem with flavor text in the first place. if there was a discrepancy that exposed somebody as an antag which would be noticeable ic, then it's an ic issue. if they're exposed by something that _nobody_ could know ic, acting on that info would be metagay and breaking the rules. i don't see the point in removing the feature instead of the metagamer, may as well remove deadchat and re-entering your body as a ghost. even that absurd example would prevent more metagaming and save admins more effort than removing flavortext.

seems like bad policy to me to change the code to accommodate those who break the rules instead of just enforcing them, but then again not my server.

**Comment by Shadowmech88** [comment]
I think the issue is people metagaming in ways that can't be reliably tracked, or reliably judged due to plausible deniability.
Sure, you can't *call them out* as an antag due to flavor text or whatever, but you could be more cautious around them than you would otherwise, in a way that can't be detected by others but still impacts the round nonetheless.

If we have the ability to patch those holes via code without any major impact to the quality of the feature, shouldn't we?

**Comment by cadaverouscoyote** [comment]
>but you could be more cautious around them than you would otherwise, in a way that can't be detected by others but still impacts the round nonetheless

that's why i drew the comparison to deadchat and ghosting for.

>If we have the ability to patch those holes via code without any major impact to the quality of the feature, shouldn't we?

yeah i'm just gettin pissed it's taking more than a month to re-add a feature removed over what's a non-issue imo, which has been mostly rectified. couldn't merge it three weeks ago then patch it up later when there's ass to be dragged, lord no.

**Comment by nuklearcellphoneg** [comment]
It was removed for pretty big metagaming issues honestly.

**Comment by Exxion** [comment]
Yeah that

This implementation doesn't have a button to change flavor text mid-round, which is why the popup for certain antag types to set their flavor text at roundstart is important.
This implementation *couldn't* really have that button, because it attaches flavor text to DNA, and multiple people can have the same DNA if there's a changeling or wacky geneticist antics.

**Comment by cadaverouscoyote** [comment]
well fuckin shit

**Comment by CodenameB** [comment]
I haven't done the wizard/ops/raiders fix because I still don't really truly understand flavor text since someone else wrote it, and I don't understand how the wizard's name popup works either. I'll get working on it some weekend, but would love if someone else could take a look.

**Comment by SonixApache** [comment]
@CodenameB conflicts

**Comment by Riekland** [comment]
wait hold on why aren't we just porting the bay flavour text that has you make different lines for different body parts that only show up if they're not covered up

**Comment by Exxion** [comment]
1. Someone would have to actually do it
2. More trouble for administrators having to administrate each piece of flavor text
3. The unfathomable amount of shitposting that would occur virtually everywhere

In that order.

**Comment by Blithering** [comment]
>not tattooing your name and blood type on each body part for easy identification

**Comment by N3X15** [comment]
As much as I hate enabling the waifushittery, please fix the conflicts and contact me so I can get a parallel MR up on the secret repo to fix the secret repo conflict.

**Comment by Exxion** [comment]
We were still waiting on something
Though I guess I should have actually hit the request changes button to make it more clear

**Comment by Riekland** [comment]
hey nexis remember when you said that you wouldn't do this kind of shit after you took your bi-weekly break for realsies this time
haha yeah

**Comment by Intigracy** [comment]
This broke SQL

**Pull request merged**